### ISSUE TITLE
performance scaling issue

### ISSUE THREAD

**Issue by linehrr** [facebook/presto#4292]
in the later release(0.125, 0.131 that we tested), we found that the performance for complex queries downgraded a lot.
looking at the CPU usage on the cloudera monitor, CPU is always low, almost 0 and the query just takes forever.

we used to have a good CPU utilization on the previous releases, like our production release: 0.112, 0.110 etc.

here is our configuration on the presto server when we were testing on 0.131:
jvm.config:
```
-server
-Xmx18G
-XX:+UseConcMarkSweepGC
-XX:+UseParNewGC
-XX:-CMSConcurrentMTEnabled
-XX:CMSInitiatingOccupancyFraction=70
-XX:+CMSParallelRemarkEnabled
-XX:+ExplicitGCInvokesConcurrent
-XX:+CMSClassUnloadingEnabled
-XX:+AggressiveOpts
-XX:OnOutOfMemoryError=kill -9 %p
-XX:PermSize=150M
-XX:MaxPermSize=150M
-XX:ReservedCodeCacheSize=150M
-Xbootclasspath/p:/opt/presto/lib/floatingdecimal-0.1.jar
```

config.properties:
```
coordinator=true
datasources=jmx,hive
http-server.http.port=8080
discovery-server.enabled=true
discovery.uri=http://0.0.0.0:8080
node-scheduler.include-coordinator=false
exchange.http-client.connect-timeout=30m
exchange.http-client.read-timeout=30m
scheduler.http-client.connect-timeout=30m
scheduler.http-client.read-timeout=30m
execute.http-client.read-timeout=10m
failure-detector.threshold=0.10
query.queue-config-file=/opt/presto/etc/queue.conf
```

**Comment by linehrr** [comment]
specifically when v0.125 and v0.131 query stuck for a long time, if we look at the threads status, 99% of the threads are been marked as BLOCKING.
so maybe there is a query planning issue in the coordinator.

**Comment by cawallin** [comment]
Can you give us some examples of queries that you're running?

**Comment by linehrr** [comment]
here is one of the queries that had slow run in new release but fast run in previous release.
table parsed_logs is partitioned by type and data is in RCFile.
parsed_logs is about 7TB of size.
and those join tables on the right side are about 1-10MB.

```
select  ppp.net,
        adc.datacenter,
        parsed_logs.siteid,
        sites.name,
        browser,
        os,
        count_if(type = 'RB' and substr(userid, 1, 1) = '_') as rb_ipuserid,
        count_if(type = 'RP' and substr(userid, 1, 1) = '_') as rp_ipuserid,
        count_if(type in ('RI') and substr(userid, 1, 1) = '_') as ri_ipuserid,
        count_if(type = 'RB') as rb_total,
        count_if(type = 'RP') as rp_total,
        count_if(type in ('RI')) as ri_total
from parsed_logs
join sites on sites.siteid = parsed_logs.siteid
join ppp on ppp.userid = sites.userid
join mappings_sitetype mt on mt.siteid = parsed_logs.siteid and mt.sitetype = 'proxy'
join datacenter_mappings adc on adc.asname = parsed_logs.asname
where dt = '2016-01-05'
        and type in ('RI','RB','RP','BRB','KRB','BKRB','BRP','KRP','BKRP')
group by ppp.net,
        adc.datacenter,
        parsed_logs.siteid,
        sites.name,
        browser,
        os
```

**Comment by KBP-TDC** [comment]
Could you please provide us with an output of EXPLAIN? Comparing the new one to the old one (before the upgrade) may reveal the issue.

**Comment by linehrr** [comment]
they do have difference.
especially at the join strategy..
new one seems to hash join and old one does not.

old explain:
```
 - Output[rtbnetwork, datacenter, siteid, name, browser, os, rb_ipcmid, rp_ipcmid, ri_ipcmid, rb_total, rp_total, ri_total] => [rtbnetwork:varchar, datacenter:varchar, siteid:varchar, name:varchar, browser:varchar, os:varchar, count_if:bigint, count_if_25:bigint, count_if_26:bigint, count_if_27:bigint, count_if_28:bigint, count_if_29:bigint] 
         rb_ipcmid := count_if                                                                                                                                                                                                                                                                                                                          
         rp_ipcmid := count_if_25                                                                                                                                                                                                                                                                                                                       
         ri_ipcmid := count_if_26                                                                                                                                                                                                                                                                                                                       
         rb_total := count_if_27                                                                                                                                                                                                                                                                                                                        
         rp_total := count_if_28                                                                                                                                                                                                                                                                                                                        
         ri_total := count_if_29                                                                                                                                                                                                                                                                                                                        
     - Exchange[GATHER] => rtbnetwork:varchar, datacenter:varchar, siteid:varchar, name:varchar, browser:varchar, os:varchar, count_if_27:bigint, count_if_26:bigint, count_if_25:bigint, count_if:bigint, count_if_29:bigint, count_if_28:bigint                                                                                                       
         - Aggregate(FINAL)[rtbnetwork, datacenter, siteid, name, browser, os] => [rtbnetwork:varchar, datacenter:varchar, siteid:varchar, name:varchar, browser:varchar, os:varchar, count_if_27:bigint, count_if_26:bigint, count_if_25:bigint, count_if:bigint, count_if_29:bigint, count_if_28:bigint]                                              
                 count_if_27 := "count_if"("count_if_83")                                                                                                                                                                                                                                                                                               
                 count_if_26 := "count_if"("count_if_82")                                                                                                                                                                                                                                                                                               
                 count_if_25 := "count_if"("count_if_81")                                                                                                                                                                                                                                                                                               
                 count_if := "count_if"("count_if_80")                                                                                                                                                                                                                                                                                                  
                 count_if_29 := "count_if"("count_if_85")                                                                                                                                                                                                                                                                                               
                 count_if_28 := "count_if"("count_if_84")                                                                                                                                                                                                                                                                                               
             - Exchange[REPARTITION] => rtbnetwork:varchar, datacenter:varchar, siteid:varchar, name:varchar, browser:varchar, os:varchar, count_if_85:bigint, count_if_84:bigint, count_if_83:bigint, count_if_82:bigint, count_if_81:bigint, count_if_80:bigint                                                                                       
                 - Aggregate(PARTIAL)[rtbnetwork, datacenter, siteid, name, browser, os] => [rtbnetwork:varchar, datacenter:varchar, siteid:varchar, name:varchar, browser:varchar, os:varchar, count_if_85:bigint, count_if_84:bigint, count_if_83:bigint, count_if_82:bigint, count_if_81:bigint, count_if_80:bigint]                                 
                         count_if_85 := "count_if"("expr_24")                                                                                                                                                                                                                                                                                           
                         count_if_84 := "count_if"("expr_23")                                                                                                                                                                                                                                                                                           
                         count_if_83 := "count_if"("expr_22")                                                                                                                                                                                                                                                                                           
                         count_if_82 := "count_if"("expr_21")                                                                                                                                                                                                                                                                                           
                         count_if_81 := "count_if"("expr_20")                                                                                                                                                                                                                                                                                           
                         count_if_80 := "count_if"("expr_19")                                                                                                                                                                                                                                                                                           
                     - Project => [rtbnetwork:varchar, datacenter:varchar, siteid:varchar, name:varchar, browser:varchar, os:varchar, expr_19:boolean, expr_20:boolean, expr_21:boolean, expr_22:boolean, expr_23:boolean, expr_24:boolean]                                                                                                             
                             expr_19 := (("type" = 'RB') AND ("substr"("cmid", 1, 1) = '_'))                                                                                                                                                                                                                                                            
                             expr_20 := (("type" = 'RP') AND ("substr"("cmid", 1, 1) = '_'))                                                                                                                                                                                                                                                            
                             expr_21 := (("type" IN ('RI')) AND ("substr"("cmid", 1, 1) = '_'))                                                                                                                                                                                                                                                         
                             expr_22 := ("type" = 'RB')                                                                                                                                                                                                                                                                                                 
                             expr_23 := ("type" = 'RP')                                                                                                                                                                                                                                                                                                 
                             expr_24 := ("type" IN ('RI'))                                                                                                                                                                                                                                                                                              
                         - InnerJoin[("asname" = "asname_10")] => [rtbnetwork:varchar, type:varchar, siteid:varchar, cmid:varchar, asname:varchar, name:varchar, browser:varchar, os:varchar, asname_10:varchar, datacenter:varchar]                                                                                                                    
                             - Project => [rtbnetwork:varchar, type:varchar, siteid:varchar, cmid:varchar, asname:varchar, name:varchar, browser:varchar, os:varchar]                                                                                                                                                                                   
                                 - InnerJoin[("siteid" = "siteid_6")] => [asname:varchar, os:varchar, rtbnetwork:varchar, browser:varchar, name:varchar, siteid:varchar, cmid:varchar, type:varchar, siteid_6:varchar]                                                                                                                                  
                                     - Project => [asname:varchar, os:varchar, rtbnetwork:varchar, browser:varchar, name:varchar, siteid:varchar, cmid:varchar, type:varchar]                                                                                                                                                                           
                                         - InnerJoin[("userid" = "userid_3")] => [asname:varchar, os:varchar, browser:varchar, name:varchar, siteid:varchar, cmid:varchar, type:varchar, userid:varchar, userid_3:varchar, rtbnetwork:varchar]                                                                                                          
                                             - Project => [asname:varchar, os:varchar, browser:varchar, name:varchar, siteid:varchar, cmid:varchar, type:varchar, userid:varchar]                                                                                                                                                                       
                                                 - InnerJoin[("siteid" = "siteid_0")] => [siteid:varchar, asname:varchar, cmid:varchar, os:varchar, browser:varchar, type:varchar, userid:varchar, siteid_0:varchar, name:varchar]                                                                                                                      
                                                     - TableScan[hive:hive:default:parsed_auditlogs, originalConstraint = (("type" IN ('RI', 'RB', 'RP', 'BRB', 'KRB', 'BKRB', 'BRP', 'KRP', 'BKRP')) AND ('2016-01-05' = "dt"))] => [siteid:varchar, asname:varchar, cmid:varchar, os:varchar, browser:varchar, type:varchar]                          
                                                             siteid := HiveColumnHandle{clientId=hive, name=siteid, ordinalPosition=0, hiveType=string, hiveColumnIndex=0, partitionKey=false}                                                                                                                                                          
                                                             asname := HiveColumnHandle{clientId=hive, name=asname, ordinalPosition=16, hiveType=string, hiveColumnIndex=16, partitionKey=false}                                                                                                                                                        
                                                             cmid := HiveColumnHandle{clientId=hive, name=cmid, ordinalPosition=19, hiveType=string, hiveColumnIndex=19, partitionKey=false}                                                                                                                                                            
                                                             os := HiveColumnHandle{clientId=hive, name=os, ordinalPosition=31, hiveType=string, hiveColumnIndex=31, partitionKey=false}                                                                                                                                                                
                                                             browser := HiveColumnHandle{clientId=hive, name=browser, ordinalPosition=32, hiveType=string, hiveColumnIndex=32, partitionKey=false}                                                                                                                                                      
                                                             type := HiveColumnHandle{clientId=hive, name=type, ordinalPosition=69, hiveType=string, hiveColumnIndex=-1, partitionKey=true}                                                                                                                                                             
                                                                 :: [[BKRB], [BKRP], [BRB], [BRP], [KRB], [KRP], [RB], [RI], [RP]]                                                                                                                                                                                                                      
                                                             HiveColumnHandle{clientId=hive, name=dt, ordinalPosition=0, hiveType=string, hiveColumnIndex=-1, partitionKey=true}                                                                                                                                                                        
                                                                 :: [[2016-01-05]]                                                                                                                                                                                                                                                                      
                                                     - Exchange[REPLICATE] => userid:varchar, siteid_0:varchar, name:varchar                                                                                                                                                                                                                            
                                                         - TableScan[hive:hive:default:sites, originalConstraint = true] => [userid:varchar, siteid_0:varchar, name:varchar]                                                                                                                                                                            
                                                                 userid := HiveColumnHandle{clientId=hive, name=userid, ordinalPosition=0, hiveType=string, hiveColumnIndex=0, partitionKey=false}                                                                                                                                                      
                                                                 siteid_0 := HiveColumnHandle{clientId=hive, name=siteid, ordinalPosition=1, hiveType=string, hiveColumnIndex=1, partitionKey=false}                                                                                                                                                    
                                                                 name := HiveColumnHandle{clientId=hive, name=name, ordinalPosition=3, hiveType=string, hiveColumnIndex=3, partitionKey=false}                                                                                                                                                          
                                             - Exchange[REPLICATE] => userid_3:varchar, rtbnetwork:varchar                                                                                                                                                                                                                                              
                                                 - TableScan[hive:hive:default:publishers, originalConstraint = true] => [userid_3:varchar, rtbnetwork:varchar]                                                                                                                                                                                         
                                                         userid_3 := HiveColumnHandle{clientId=hive, name=userid, ordinalPosition=0, hiveType=string, hiveColumnIndex=0, partitionKey=false}                                                                                                                                                            
                                                         rtbnetwork := HiveColumnHandle{clientId=hive, name=rtbnetwork, ordinalPosition=1, hiveType=string, hiveColumnIndex=1, partitionKey=false}                                                                                                                                                      
                                     - Exchange[REPLICATE] => siteid_6:varchar                                                                                                                                                                                                                                                                          
                                         - Project => [siteid_6:varchar]                                                                                                                                                                                                                                                                                
                                             - Filter[("sitetype" = 'proxy')] => [siteid_6:varchar, sitetype:varchar]                                                                                                                                                                                                                                   
                                                 - TableScan[hive:hive:default:mappings_sitetype, originalConstraint = ('proxy' = "sitetype")] => [siteid_6:varchar, sitetype:varchar]                                                                                                                                                                  
                                                         siteid_6 := HiveColumnHandle{clientId=hive, name=siteid, ordinalPosition=0, hiveType=string, hiveColumnIndex=0, partitionKey=false}                                                                                                                                                            
                                                         sitetype := HiveColumnHandle{clientId=hive, name=sitetype, ordinalPosition=1, hiveType=string, hiveColumnIndex=1, partitionKey=false}                                                                                                                                                          
                             - Exchange[REPLICATE] => asname_10:varchar, datacenter:varchar                                                                                                                                                                                                                                                             
                                 - TableScan[hive:hive:default:as_datacenter_mappings, originalConstraint = true] => [asname_10:varchar, datacenter:varchar]                                                                                                                                                                                            
                                         asname_10 := HiveColumnHandle{clientId=hive, name=asname, ordinalPosition=0, hiveType=string, hiveColumnIndex=0, partitionKey=false}                                                                                                                                                                           
                                         datacenter := HiveColumnHandle{clientId=hive, name=datacenter, ordinalPosition=1, hiveType=string, hiveColumnIndex=1, partitionKey=false}
```

new explain:
```
 - Output[rtbnetwork, datacenter, siteid, name, browser, os, rb_ipcmid, rp_ipcmid, ri_ipcmid, rb_total, rp_total, ri_total] => [rtbnetwork:varchar, datacenter:varchar, siteid:varchar, name:varchar, browser:varchar, os:varchar, count_if:bigint, count_if_25:bigint, count_if_26:bigint, count_if_27:bigint, count_if_28:bigint, count_if_29:bigint]                                                                                                                                                                      
         rb_ipcmid := count_if                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               
         rp_ipcmid := count_if_25                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            
         ri_ipcmid := count_if_26                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            
         rb_total := count_if_27                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             
         rp_total := count_if_28                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             
         ri_total := count_if_29                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             
     - Exchange[GATHER] => rtbnetwork:varchar, datacenter:varchar, siteid:varchar, name:varchar, browser:varchar, os:varchar, count_if_27:bigint, count_if_26:bigint, count_if_25:bigint, count_if:bigint, count_if_29:bigint, count_if_28:bigint                                                                                                                                                                                                                                                                            
         - Aggregate(FINAL)[rtbnetwork, datacenter, siteid, name, browser, os] => [rtbnetwork:varchar, datacenter:varchar, siteid:varchar, name:varchar, browser:varchar, os:varchar, $hashvalue_81:bigint, count_if_27:bigint, count_if_26:bigint, count_if_25:bigint, count_if:bigint, count_if_29:bigint, count_if_28:bigint]                                                                                                                                                                                             
                 count_if_27 := "count_if"("count_if_85")                                                                                                                                                                                                                                                                                                                                                                                                                                                                    
                 count_if_26 := "count_if"("count_if_84")                                                                                                                                                                                                                                                                                                                                                                                                                                                                    
                 count_if_25 := "count_if"("count_if_83")                                                                                                                                                                                                                                                                                                                                                                                                                                                                    
                 count_if := "count_if"("count_if_82")                                                                                                                                                                                                                                                                                                                                                                                                                                                                       
                 count_if_29 := "count_if"("count_if_87")                                                                                                                                                                                                                                                                                                                                                                                                                                                                    
                 count_if_28 := "count_if"("count_if_86")                                                                                                                                                                                                                                                                                                                                                                                                                                                                    
             - Exchange[REPARTITION] => rtbnetwork:varchar, datacenter:varchar, siteid:varchar, name:varchar, browser:varchar, os:varchar, $hashvalue_81:bigint, count_if_87:bigint, count_if_86:bigint, count_if_85:bigint, count_if_84:bigint, count_if_83:bigint, count_if_82:bigint                                                                                                                                                                                                                                      
                 - Aggregate(PARTIAL)[rtbnetwork, datacenter, siteid, name, browser, os] => [rtbnetwork:varchar, datacenter:varchar, siteid:varchar, name:varchar, browser:varchar, os:varchar, $hashvalue_81:bigint, count_if_87:bigint, count_if_86:bigint, count_if_85:bigint, count_if_84:bigint, count_if_83:bigint, count_if_82:bigint]                                                                                                                                                                                
                         count_if_87 := "count_if"("expr_24")                                                                                                                                                                                                                                                                                                                                                                                                                                                                
                         count_if_86 := "count_if"("expr_23")                                                                                                                                                                                                                                                                                                                                                                                                                                                                
                         count_if_85 := "count_if"("expr_22")                                                                                                                                                                                                                                                                                                                                                                                                                                                                
                         count_if_84 := "count_if"("expr_21")                                                                                                                                                                                                                                                                                                                                                                                                                                                                
                         count_if_83 := "count_if"("expr_20")                                                                                                                                                                                                                                                                                                                                                                                                                                                                
                         count_if_82 := "count_if"("expr_19")                                                                                                                                                                                                                                                                                                                                                                                                                                                                
                     - Project => [rtbnetwork:varchar, datacenter:varchar, siteid:varchar, name:varchar, browser:varchar, os:varchar, expr_19:boolean, expr_20:boolean, expr_21:boolean, expr_22:boolean, expr_23:boolean, expr_24:boolean, $hashvalue_81:bigint]                                                                                                                                                                                                                                                            
                             expr_19 := (("type" = 'RB') AND ("substr"("cmid", 1, 1) = '_'))                                                                                                                                                                                                                                                                                                                                                                                                                                 
                             expr_20 := (("type" = 'RP') AND ("substr"("cmid", 1, 1) = '_'))                                                                                                                                                                                                                                                                                                                                                                                                                                 
                             expr_21 := (("type" IN ('RI')) AND ("substr"("cmid", 1, 1) = '_'))                                                                                                                                                                                                                                                                                                                                                                                                                              
                             expr_22 := ("type" = 'RB')                                                                                                                                                                                                                                                                                                                                                                                                                                                                      
                             expr_23 := ("type" = 'RP')                                                                                                                                                                                                                                                                                                                                                                                                                                                                      
                             expr_24 := ("type" IN ('RI'))                                                                                                                                                                                                                                                                                                                                                                                                                                                                   
                             $hashvalue_81 := "combine_hash"("combine_hash"("combine_hash"("combine_hash"("combine_hash"("combine_hash"(0, COALESCE("$operator$hash_code"("rtbnetwork"), 0)), COALESCE("$operator$hash_code"("datacenter"), 0)), COALESCE("$operator$hash_code"("siteid"), 0)), COALESCE("$operator$hash_code"("name"), 0)), COALESCE("$operator$hash_code"("browser"), 0)), COALESCE("$operator$hash_code"("os"), 0))                                                                                       
                         - InnerJoin[("$hashvalue_105" = "combine_hash_106") AND ("combine_hash_107" = "$hashvalue_108") AND ("asname_109" = "asname_110")] => [asname:varchar, $hashvalue_79:bigint, os:varchar, rtbnetwork:varchar, browser:varchar, name:varchar, siteid:varchar, cmid:varchar, type:varchar, $hashvalue_105:bigint, combine_hash_107:bigint, asname_109:varchar, asname_10:varchar, $hashvalue_80:bigint, datacenter:varchar, combine_hash_106:bigint, $hashvalue_108:bigint, asname_110:varchar]        
                             - Exchange[REPARTITION] => asname:varchar, $hashvalue_79:bigint, os:varchar, rtbnetwork:varchar, browser:varchar, name:varchar, siteid:varchar, cmid:varchar, type:varchar, $hashvalue_105:bigint, combine_hash_107:bigint, asname_109:varchar                                                                                                                                                                                                                                                  
                                 - Project => [asname:varchar, $hashvalue_79:bigint, os:varchar, rtbnetwork:varchar, browser:varchar, name:varchar, siteid:varchar, cmid:varchar, type:varchar]                                                                                                                                                                                                                                                                                                                              
                                         $hashvalue_79 := "combine_hash"(0, COALESCE("$operator$hash_code"("asname"), 0))                                                                                                                                                                                                                                                                                                                                                                                                    
                                     - InnerJoin[("$hashvalue_99" = "combine_hash_100") AND ("siteid_101" = "siteid_102") AND ("combine_hash_103" = "$hashvalue_104")] => [siteid:varchar, $hashvalue_77:bigint, asname:varchar, os:varchar, rtbnetwork:varchar, browser:varchar, name:varchar, cmid:varchar, type:varchar, $hashvalue_99:bigint, siteid_101:varchar, combine_hash_103:bigint, siteid_5:varchar, $hashvalue_78:bigint, combine_hash_100:bigint, siteid_102:varchar, $hashvalue_104:bigint]                   
                                         - Exchange[REPARTITION] => siteid:varchar, $hashvalue_77:bigint, asname:varchar, os:varchar, rtbnetwork:varchar, browser:varchar, name:varchar, cmid:varchar, type:varchar, $hashvalue_99:bigint, siteid_101:varchar, combine_hash_103:bigint                                                                                                                                                                                                                                       
                                             - Project => [siteid:varchar, $hashvalue_77:bigint, asname:varchar, os:varchar, rtbnetwork:varchar, browser:varchar, name:varchar, cmid:varchar, type:varchar]                                                                                                                                                                                                                                                                                                                  
                                                     $hashvalue_77 := "combine_hash"(0, COALESCE("$operator$hash_code"("siteid"), 0))                                                                                                                                                                                                                                                                                                                                                                                        
                                                 - InnerJoin[("combine_hash_93" = "$hashvalue_94") AND ("userid_95" = "userid_96") AND ("$hashvalue_97" = "combine_hash_98")] => [userid:varchar, $hashvalue_75:bigint, asname:varchar, os:varchar, browser:varchar, name:varchar, siteid:varchar, cmid:varchar, type:varchar, combine_hash_93:bigint, userid_95:varchar, $hashvalue_97:bigint, userid_2:varchar, $hashvalue_76:bigint, rtbnetwork:varchar, $hashvalue_94:bigint, userid_96:varchar, combine_hash_98:bigint] 
                                                     - Exchange[REPARTITION] => userid:varchar, $hashvalue_75:bigint, asname:varchar, os:varchar, browser:varchar, name:varchar, siteid:varchar, cmid:varchar, type:varchar, combine_hash_93:bigint, userid_95:varchar, $hashvalue_97:bigint                                                                                                                                                                                                                                 
                                                         - Project => [userid:varchar, $hashvalue_75:bigint, asname:varchar, os:varchar, browser:varchar, name:varchar, siteid:varchar, cmid:varchar, type:varchar]                                                                                                                                                                                                                                                                                                          
                                                                 $hashvalue_75 := "combine_hash"(0, COALESCE("$operator$hash_code"("userid"), 0))                                                                                                                                                                                                                                                                                                                                                                            
                                                             - InnerJoin[("$hashvalue_88" = "combine_hash") AND ("combine_hash_89" = "$hashvalue_90") AND ("siteid_91" = "siteid_92")] => [siteid:varchar, $hashvalue:bigint, asname:varchar, os:varchar, browser:varchar, cmid:varchar, type:varchar, $hashvalue_88:bigint, combine_hash_89:bigint, siteid_91:varchar, siteid_0:varchar, $hashvalue_74:bigint, name:varchar, userid:varchar, combine_hash:bigint, $hashvalue_90:bigint, siteid_92:varchar]                  
                                                                 - Exchange[REPARTITION] => siteid:varchar, $hashvalue:bigint, asname:varchar, os:varchar, browser:varchar, cmid:varchar, type:varchar, $hashvalue_88:bigint, combine_hash_89:bigint, siteid_91:varchar                                                                                                                                                                                                                                                      
                                                                     - Project => [siteid:varchar, $hashvalue:bigint, asname:varchar, os:varchar, browser:varchar, cmid:varchar, type:varchar]                                                                                                                                                                                                                                                                                                                               
                                                                             $hashvalue := "combine_hash"(0, COALESCE("$operator$hash_code"("siteid"), 0))                                                                                                                                                                                                                                                                                                                                                                   
                                                                         - TableScan[hive:hive:default:parsed_auditlogs, originalConstraint = (("type" IN ('RI', 'RB', 'RP', 'BRB', 'KRB', 'BKRB', 'BRP', 'KRP', 'BKRP')) AND ('2016-01-05' = "dt"))] => [siteid:varchar, asname:varchar, cmid:varchar, os:varchar, browser:varchar, type:varchar]                                                                                                                                                                           
                                                                                 LAYOUT: hive                                                                                                                                                                                                                                                                                                                                                                                                                                
                                                                                 :: NONE                                                                                                                                                                                                                                                                                                                                                                                                                                     
                                                                 - Exchange[REPARTITION] => siteid_0:varchar, $hashvalue_74:bigint, name:varchar, userid:varchar, combine_hash:bigint, $hashvalue_90:bigint, siteid_92:varchar                                                                                                                                                                                                                                                                                               
                                                                     - Project => [siteid_0:varchar, $hashvalue_74:bigint, name:varchar, userid:varchar]                                                                                                                                                                                                                                                                                                                                                                     
                                                                             $hashvalue_74 := "combine_hash"(0, COALESCE("$operator$hash_code"("siteid_0"), 0))                                                                                                                                                                                                                                                                                                                                                              
                                                                         - Filter[false] => [userid:varchar, siteid_0:varchar, name:varchar]                                                                                                                                                                                                                                                                                                                                                                                 
                                                                             - TableScan[hive:hive:default:sites, originalConstraint = true] => [userid:varchar, siteid_0:varchar, name:varchar]                                                                                                                                                                                                                                                                                                                             
                                                                                     LAYOUT: hive                                                                                                                                                                                                                                                                                                                                                                                                                            
                                                                                     userid := HiveColumnHandle{clientId=hive, name=userid, hiveType=string, hiveColumnIndex=0, partitionKey=false}                                                                                                                                                                                                                                                                                                                          
                                                                                     siteid_0 := HiveColumnHandle{clientId=hive, name=siteid, hiveType=string, hiveColumnIndex=1, partitionKey=false}                                                                                                                                                                                                                                                                                                                        
                                                                                     name := HiveColumnHandle{clientId=hive, name=name, hiveType=string, hiveColumnIndex=3, partitionKey=false}                                                                                                                                                                                                                                                                                                                              
                                                     - Exchange[REPARTITION] => userid_2:varchar, $hashvalue_76:bigint, rtbnetwork:varchar, $hashvalue_94:bigint, userid_96:varchar, combine_hash_98:bigint                                                                                                                                                                                                                                                                                                                  
                                                         - Project => [userid_2:varchar, $hashvalue_76:bigint, rtbnetwork:varchar]                                                                                                                                                                                                                                                                                                                                                                                           
                                                                 $hashvalue_76 := "combine_hash"(0, COALESCE("$operator$hash_code"("userid_2"), 0))                                                                                                                                                                                                                                                                                                                                                                          
                                                             - Filter[false] => [userid_2:varchar, rtbnetwork:varchar]                                                                                                                                                                                                                                                                                                                                                                                                       
                                                                 - TableScan[hive:hive:default:publishers, originalConstraint = true] => [userid_2:varchar, rtbnetwork:varchar]                                                                                                                                                                                                                                                                                                                                              
                                                                         LAYOUT: hive                                                                                                                                                                                                                                                                                                                                                                                                                                        
                                                                         userid_2 := HiveColumnHandle{clientId=hive, name=userid, hiveType=string, hiveColumnIndex=0, partitionKey=false}                                                                                                                                                                                                                                                                                                                                    
                                                                         rtbnetwork := HiveColumnHandle{clientId=hive, name=rtbnetwork, hiveType=string, hiveColumnIndex=1, partitionKey=false}                                                                                                                                                                                                                                                                                                                              
                                         - Exchange[REPARTITION] => siteid_5:varchar, $hashvalue_78:bigint, combine_hash_100:bigint, siteid_102:varchar, $hashvalue_104:bigint                                                                                                                                                                                                                                                                                                                                               
                                             - Project => [siteid_5:varchar, $hashvalue_78:bigint]                                                                                                                                                                                                                                                                                                                                                                                                                           
                                                     $hashvalue_78 := "combine_hash"(0, COALESCE("$operator$hash_code"("siteid_5"), 0))                                                                                                                                                                                                                                                                                                                                                                                      
                                                 - Filter[false] => [siteid_5:varchar]                                                                                                                                                                                                                                                                                                                                                                                                                                       
                                                     - TableScan[hive:hive:default:mappings_sitetype, originalConstraint = ('proxy' = "sitetype")] => [siteid_5:varchar]                                                                                                                                                                                                                                                                                                                                                     
                                                             LAYOUT: hive                                                                                                                                                                                                                                                                                                                                                                                                                                                    
                                                             siteid_5 := HiveColumnHandle{clientId=hive, name=siteid, hiveType=string, hiveColumnIndex=0, partitionKey=false}                                                                                                                                                                                                                                                                                                                                                
                             - Exchange[REPARTITION] => asname_10:varchar, $hashvalue_80:bigint, datacenter:varchar, combine_hash_106:bigint, $hashvalue_108:bigint, asname_110:varchar                                                                                                                                                                                                                                                                                                                                      
                                 - Project => [asname_10:varchar, $hashvalue_80:bigint, datacenter:varchar]                                                                                                                                                                                                                                                                                                                                                                                                                  
                                         $hashvalue_80 := "combine_hash"(0, COALESCE("$operator$hash_code"("asname_10"), 0))                                                                                                                                                                                                                                                                                                                                                                                                 
                                     - Filter[false] => [asname_10:varchar, datacenter:varchar]                                                                                                                                                                                                                                                                                                                                                                                                                              
                                         - TableScan[hive:hive:default:as_datacenter_mappings, originalConstraint = true] => [asname_10:varchar, datacenter:varchar]                                                                                                                                                                                                                                                                                                                                                         
                                                 LAYOUT: hive                                                                                                                                                                                                                                                                                                                                                                                                                                                                
                                                 asname_10 := HiveColumnHandle{clientId=hive, name=asname, hiveType=string, hiveColumnIndex=0, partitionKey=false}                                                                                                                                                                                                                                                                                                                                                           
                                                 datacenter := HiveColumnHandle{clientId=hive, name=datacenter, hiveType=string, hiveColumnIndex=1, partitionKey=false}
```

**Comment by pnowojski** [comment]
@linehrr from Presto 0.116 `distributed-joins-enabled` option is set by default to `true`. We had very similar (or maybe exactly the same) issue as you have described and it was solved by setting this option back to `false`.

Probably the problem is that this queries requires lots of RAM to run using distributed joins and Presto coordinator is not willing to assign all of the available memory to single query at once and is constantly blocking the query. 

https://prestodb.io/docs/current/release/release-0.116.html
https://prestodb.io/docs/current/admin/tuning.html

**Comment by linehrr** [comment]
thanks @pnowojski , your answer solved the problem.
the query speeds up to normal.
however we've noticed that presto crashes more in the newer release?
have you encountered the similar issues?

err msg said:
```
com.facebook.presto.spi.PrestoException: Encountered too many errors talking to a worker node. The node may have crashed or be under too much load. This is probably a transient issue, so please retry your query in a few minutes. (http://172.16.20.44:8080/v1/task/20160108_111444_00033_aticz.2.81 - 23 failures, time since last success 120.97s)
	at com.facebook.presto.server.HttpRemoteTask$RequestErrorTracker.requestFailed(HttpRemoteTask.java:892)
	at com.facebook.presto.server.HttpRemoteTask$ContinuousTaskInfoFetcher.failed(HttpRemoteTask.java:735)
	at com.facebook.presto.server.HttpRemoteTask$SimpleHttpResponseHandler.onSuccess(HttpRemoteTask.java:789)
	at com.facebook.presto.server.HttpRemoteTask$SimpleHttpResponseHandler.onSuccess(HttpRemoteTask.java:768)
	at com.google.common.util.concurrent.Futures$6.run(Futures.java:1319)
	at io.airlift.concurrent.BoundedExecutor.executeOrMerge(BoundedExecutor.java:69)
	at io.airlift.concurrent.BoundedExecutor.access$000(BoundedExecutor.java:28)
	at io.airlift.concurrent.BoundedExecutor$1.run(BoundedExecutor.java:40)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
	Suppressed: com.facebook.presto.server.HttpRemoteTask.ServiceUnavailableException: Server returned SERVICE_UNAVAILABLE: http://172.16.20.44:8080/v1/task/20160108_111444_00033_aticz.2.81?summarize
		... 9 more
	Suppressed: com.facebook.presto.server.HttpRemoteTask.ServiceUnavailableException: Server returned SERVICE_UNAVAILABLE: http://172.16.20.44:8080/v1/task/20160108_111444_00033_aticz.2.81?summarize
		... 9 more
	Suppressed: com.facebook.presto.server.HttpRemoteTask.ServiceUnavailableException: Server returned SERVICE_UNAVAILABLE: http://172.16.20.44:8080/v1/task/20160108_111444_00033_aticz.2.81?summarize
		... 9 more
	Suppressed: com.facebook.presto.server.HttpRemoteTask.ServiceUnavailableException: Server returned SERVICE_UNAVAILABLE: http://172.16.20.44:8080/v1/task/20160108_111444_00033_aticz.2.81?summarize
		... 9 more
	Suppressed: com.facebook.presto.server.HttpRemoteTask.ServiceUnavailableException: Server returned SERVICE_UNAVAILABLE: http://172.16.20.44:8080/v1/task/20160108_111444_00033_aticz.2.81?summarize
		... 9 more
	Suppressed: com.facebook.presto.server.HttpRemoteTask.ServiceUnavailableException: Server returned SERVICE_UNAVAILABLE: http://172.16.20.44:8080/v1/task/20160108_111444_00033_aticz.2.81?summarize
		... 9 more
	Suppressed: com.facebook.presto.server.HttpRemoteTask.ServiceUnavailableException: Server returned SERVICE_UNAVAILABLE: http://172.16.20.44:8080/v1/task/20160108_111444_00033_aticz.2.81?summarize
		... 9 more
	Suppressed: com.facebook.presto.server.HttpRemoteTask.ServiceUnavailableException: Server returned SERVICE_UNAVAILABLE: http://172.16.20.44:8080/v1/task/20160108_111444_00033_aticz.2.81?summarize
		... 9 more
	Suppressed: com.facebook.presto.server.HttpRemoteTask.ServiceUnavailableException: Server returned SERVICE_UNAVAILABLE: http://172.16.20.44:8080/v1/task/20160108_111444_00033_aticz.2.81?summarize
		... 9 more
	Suppressed: com.facebook.presto.server.HttpRemoteTask.ServiceUnavailableException: Server returned SERVICE_UNAVAILABLE: http://172.16.20.44:8080/v1/task/20160108_111444_00033_aticz.2.81?summarize
```

**Comment by KBP-TDC** [comment]
> however we've noticed that presto crashes more in the newer release?
Not really. Any chance you have:
- more data than previously
- more concurrent queries
or anything else along these lines?

Are you observing very heavy CPU load on your nodes?
Is it possible for you to increase the amount of RAM allocated to Presto?

**Comment by linehrr** [comment]
hi @KBP-TDC ,
we are observing heavy load on the CPU which is a good thing meaning CPU utilization is high.
we have limited RAM config some servers only have 16GB RAM, some servers have 32GB.
increase the RAM won't be done easily because we have hundreds of machines.
so is there any way to let Presto run at the low RAM config without crashing?

**Comment by KBP-TDC** [comment]
One way to avoid overloading Presto is to limit the number of concurrent queries by configuring Presto Queues (https://prestodb.io/docs/current/admin/queue.html). Please let us know if this helps in any way.

**Comment by miniway** [comment]
We've also noticed much more PAGE_TRANSPORT_TIMEOUT and TOO_MANY_REQUEST_FAILED after upgrading 0.128 to 0.131. Once or twice in a week before, but one or twice in an hour now. We also feel like having more frequent and longer GC time. So we have to rollback to 0.128.

**Comment by linehrr** [comment]
exactly agreed with @miniway .
many more errors seem to happen only in the recent release.
old releases seem to be much more robust.

@KBP-TDC 
we actually had the queue config, max allowed running jobs in parallel is only 2.
and also, whenever Presto crashed due to OOM.
kernel msg says:
```
7428 total pagecache pages
4876 pages in swap cache
Swap cache stats: add 740271741, delete 740266865, find 341028972/399568009
Free swap  = 0kB
Total swap = 5119992kB
4194303 pages RAM
115372 pages reserved
7038 pages shared
4039707 pages non-shared
[ pid ]   uid  tgid total_vm      rss cpu oom_adj oom_score_adj name
[  592]     0   592     2744        0  10     -17         -1000 udevd
[ 6287]     0  6287    23300       46   0     -17         -1000 auditd
[ 6312]     0  6312    62465      534   1       0             0 rsyslogd
[ 6341]     0  6341     2705       49   0       0             0 irqbalance
[ 6438]    81  6438     5351       44   9       0             0 dbus-daemon
[ 6454]     0  6454    47337        1   1       0             0 cupsd
[ 6465]     0  6465     1020        0   3       0             0 acpid
[ 6474]    68  6474     9516      169   3       0             0 hald
[ 6475]     0  6475     5082        1   3       0             0 hald-runner
[ 6507]     0  6507     5612        1   8       0             0 hald-addon-inpu
[ 6527]    68  6527     4484        1   1       0             0 hald-addon-acpi
[ 6646]     0  6646    16654        9   8     -17         -1000 sshd
[ 6654]    38  6654     7680       37  14       0             0 ntpd
[ 6730]     0  6730    20324       30   5       0             0 master
[ 6750]    89  6750    20387       28   1       0             0 qmgr
[ 6754]     0  6754    27580        1   7       0             0 abrtd
[ 6765]     0  6765    44289      144   0       0             0 tuned
[ 6777]     0  6777   708224   158547   0       0             0 python
[ 6802]     0  6802    29325       27   8       0             0 crond
[ 6813]     0  6813     5385        1  14       0             0 atd
[ 6839]     0  6839     1016        1   6       0             0 mingetty
[ 6841]     0  6841     1016        1  11       0             0 mingetty
[ 6843]     0  6843     1016        1  12       0             0 mingetty
[ 6845]     0  6845     1016        1   4       0             0 mingetty
[ 6847]     0  6847     1016        1  10       0             0 mingetty
[ 6850]     0  6850     1016        1   2       0             0 mingetty
[ 6854]     0  6854    49969      946   0       0             0 python
[ 6855]     0  6855    48443        6   5       0             0 python
[ 7099]   493  7099   474029   203885  10       0             0 java
[ 5467]     0  5467    31945       11   1       0             0 presto_daemon.p
[20104]     0 20104     2664        0   2     -17         -1000 udevd
[20171]     0 20171     2664        0   1     -17         -1000 udevd
[16878]   496 16878   461635    93380   1       0             0 java
[28337]     0 28337    40937       13   3       0             0 su
[28338]   501 28338  3706155  1064058   0       0             0 presto-server
[25401]    89 25401    20344       29   1       0             0 pickup
[ 5005]   493  5005   407745   105060  14       0             0 java
[ 5051]   493  5051   403275   103101   4       0             0 java
[ 5052]   493  5052   406735    93116   2       0             0 java
[ 5065]   493  5065   403973    99414  14       0             0 java
[ 5068]   493  5068   407553    88042   2       0             0 java
[ 5482]   493  5482   407094   102970   0       0             0 java
[ 5483]   493  5483   406331    95998   0       0             0 java
[ 5488]   493  5488   404333    92528   8       0             0 java
[ 7160]   493  7160   402469    72393  13       0             0 java
[ 7164]   493  7164   399365    80417   9       0             0 java
[ 7230]   493  7230   400011    63120  15       0             0 java
[ 7232]   493  7232   401949   114721   9       0             0 java
[ 7236]   493  7236   402932   101443  15       0             0 java
[ 7456]   493  7456   399433    49585  10       0             0 java
[ 7459]   493  7459   402014   102127   4       0             0 java
[ 7467]   493  7467   399601    69812   1       0             0 java
[ 8050]   493  8050   105175   103732   0       0             0 mapper_go
[ 8070]   493  8070   162690   159875   4       0             0 mapper_go
[ 8076]   493  8076   107687   106510   0       0             0 mapper_go
[ 8102]   493  8102   116570   115311   0       0             0 mapper_go
[ 8104]   493  8104   157181   155619   3       0             0 mapper_go
[ 8130]   493  8130   110410   109360   9       0             0 mapper_go
[ 8144]   493  8144   158815   155714   2       0             0 mapper_go
[ 8160]   493  8160   114118   112937  11       0             0 mapper_go
[ 8528]   496  8528    25281       72   3       0             0 du
[ 8613]     0  8613     4346       59  12       0             0 anacron
Out of memory: Kill process 28338 (presto-server) score 249 or sacrifice child
Killed process 28338, UID 501, (presto-server) total-vm:14824620kB, anon-rss:4256056kB, file-rss:176kB
```
so looks like the kernel killed the process.
and actually by looking at the dstat, I don't see mem was totally used up.

**Comment by cberner** [comment]
I'm investigating a possibly related issue, where the http server in the worker takes 30+ secs to send responses

**Comment by dain** [comment]
I see a number of changes or problems in the explain plans:

1) As pointed out by others, the old plan uses distributed joins and the newer one use broadcast joins

2) The new plan contains many occurrences of "Filter[false]”, and since these are all inner joins, this query should return no results
There should be an optimizer to fold "Filter[false]” to an empty VALUES node, and another optimizer to fold an inner join over an empty values node to another empty VALUES node.

3) The new plan uses optimize_hash_generation and the old one doesn’t (this is why you are getting “combine_hash” projections)
This optimization improves performance by pushing down hash code calculations to the bottom of the plan where we have more parallelism, and it reuses the hash calculations so we only have to compute them once.  

4) The pushdown optimizers don’t handle the hash calculations correctly
We are getting bogus join criteria like ("$hashvalue_99" = "combine_hash_100") AND ("siteid_101" = "siteid_102") AND ("combine_hash_103" = "$hashvalue_104”).  In this example all of the symbols containing “hash” are calculated from “siteid”, so only the middle criteria is needed.  This will result in much larger and more expensive hash tables and more data moving through exchanges.  (this is a known issue)

5) (Not from the plans)  Timeouts in newer releases actually work.
In a release (in Q4 I think), I fixed the failure code so that timeouts are actually respected.  This can cause more failures in more loaded systems, so you may need to increase the timeouts.  BTW, if the system is 100% loaded during these queries, extending the timeouts won’t help and you will need to add more resources.


Here is my (hand) simplifications of the explain plans:

old explain:

 - InnerJoin[("asname" = "asname_10")]
     - InnerJoin[("siteid" = "siteid_6")]
         - InnerJoin[("userid" = "userid_3")]
             - InnerJoin[("siteid" = "siteid_0")]
                 - TableScan[hive:hive:default:parsed_auditlogs]
                 - Exchange[REPLICATE]
                     - TableScan[hive:hive:default:sites, originalConstraint = true]
             - Exchange[REPLICATE]
                 - TableScan[hive:hive:default:publishers, originalConstraint = true]
         - Exchange[REPLICATE]
             - Filter[("sitetype" = 'proxy')]
                 - TableScan[hive:hive:default:mappings_sitetype]
     - Exchange[REPLICATE]
         - TableScan[hive:hive:default:as_datacenter_mappings, originalConstraint = true]

new explain:

 - InnerJoin[("$hashvalue_105" = "combine_hash_106") AND ("combine_hash_107" = "$hashvalue_108") AND ("asname_109" = "asname_110")]
     - Exchange[REPARTITION]
         - Project [$hashvalue_79 := "combine_hash"(0, COALESCE("$operator$hash_code"("asname"), 0))]
             - InnerJoin[("$hashvalue_99" = "combine_hash_100") AND ("siteid_101" = "siteid_102") AND ("combine_hash_103" = "$hashvalue_104")]
                 - Exchange[REPARTITION]
                     - Project [$hashvalue_77 := "combine_hash"(0, COALESCE("$operator$hash_code"("siteid"), 0))]
                         - InnerJoin[("combine_hash_93" = "$hashvalue_94") AND ("userid_95" = "userid_96") AND ("$hashvalue_97" = "combine_hash_98")]
                             - Exchange[REPARTITION]
                                 - Project [$hashvalue_75 := "combine_hash"(0, COALESCE("$operator$hash_code"("userid"), 0))]
                                     - InnerJoin[("$hashvalue_88" = "combine_hash") AND ("combine_hash_89" = "$hashvalue_90") AND ("siteid_91" = "siteid_92")]
                                         - Exchange[REPARTITION]
                                             - Project [$hashvalue := "combine_hash"(0, COALESCE("$operator$hash_code"("siteid"), 0))]
                                                 - TableScan[hive:hive:default:parsed_auditlogs, originalConstraint = (("type" IN ('RI', 'RB', 'RP', 'BRB', 'KRB', 'BKRB', 'BRP', 'KRP', 'BKRP')) AND ('2016-01-05' = "dt"))]
                                         - Exchange[REPARTITION]
                                             - Project[$hashvalue_74 := "combine_hash"(0, COALESCE("$operator$hash_code"("siteid_0"), 0))]
                                                 - Filter[false]
                                                     - TableScan[hive:hive:default:sites, originalConstraint = true]
                             - Exchange[REPARTITION]
                                 - Project[$hashvalue_76 := "combine_hash"(0, COALESCE("$operator$hash_code"("userid_2"), 0))]
                                     - Filter[false]
                                         - TableScan[hive:hive:default:publishers, originalConstraint = true]
                 - Exchange[REPARTITION]
                     - Project [$hashvalue_78 := "combine_hash"(0, COALESCE("$operator$hash_code"("siteid_5"), 0))]
                         - Filter[false]
                             - TableScan[hive:hive:default:mappings_sitetype, originalConstraint = ('proxy' = "sitetype")]
     - Exchange[REPARTITION]
         - Project
                 $hashvalue_80 := "combine_hash"(0, COALESCE("$operator$hash_code"("asname_10"), 0))
             - Filter[false]
                 - TableScan[hive:hive:default:as_datacenter_mappings, originalConstraint = true]

**Comment by linehrr** [comment]
@dain:
thanks for the explanation.
> so you may need to increase the timeouts

what parameter was this exactly? 
we have a bunch of timeouts right? so which one specifically?

**Comment by miniway** [comment]
@linehrr 
```
exchange.http-client.read-timeout=10s
exchange.min-error-duration=60s (default)
```
seems to be the related parameters.

For us, when timeout occurred, even 204 status response took more than 30secs. There's no full GC at all. 

```
2016-01-16T01:02:34.108Z        10.166.179.75   GET     /v1/task/20160116_010154_00044_6nkpm.96.2/results/20160116_010154_00044_6nkpm.95.10/0   null    null    204     0       0       33169   null
2016-01-16T01:02:34.187Z        10.166.179.75   GET     /v1/task/20160116_010154_00044_6nkpm.94.2/results/20160116_010154_00044_6nkpm.93.9/0    null    null    204     0       0       33091   null
2016-01-16T01:02:34.189Z        10.166.179.75   GET     /v1/task/20160116_010154_00044_6nkpm.104.2/results/20160116_010154_00044_6nkpm.103.3/0  null    null    204     0       0       33089   null
2016-01-16T01:02:34.194Z        10.166.179.75   GET     /v1/task/20160116_010154_00044_6nkpm.64.2/results/20160116_010154_00044_6nkpm.63.3/0    null    null    204     0       0       33084   null
...
2016-01-16T01:02:34.293Z        10.65.156.220   GET     /v1/task/20160116_010154_00044_6nkpm.738.2/results/20160116_010154_00044_6nkpm.737.0/0  null    null    204     0       0       33113   null
2016-01-16T01:02:34.318Z        10.33.146.205   GET     /v1/task/20160116_010154_00044_6nkpm.450.1/results/20160116_010154_00044_6nkpm.449.4/0  null    null    204     0       0       33093   null
2016-01-16T01:02:34.292Z        10.33.146.205   GET     /v1/task/20160116_010154_00044_6nkpm.334.1/results/20160116_010154_00044_6nkpm.333.2/0  null    null    204     0       0       33143   null
```

**Comment by linehrr** [comment]
@miniway thank you so much.
we are going to test to tweak this parameter to see if we get better success rate.
I will post my findings back to this thread.

**Comment by nezihyigitbasi** [comment]
@linehrr I think #4182 is related and in that issue there is a number of recommendations to get rid of this "Encountered too many errors ..." problem. Please give them a try too and let us know.

**Comment by linehrr** [comment]
@nezihyigitbasi thank you, that's very helpful.
however, do you know any place that has a detailed explain on each of the presto config parameters?
especially for those ones in config.properties?

**Comment by nezihyigitbasi** [comment]
@linehrr I usually go back to the source code to understand what a parameter is doing (unless it's documented in presto docs)

**Comment by linehrr** [comment]
@nezihyigitbasi 
I see, Presto doc about those parameters is pretty poor.
and lots of those are not even mentioned in the doc...
I think we need a better doc for Presto, so that ppl can fully take advantage of tuning.

**Comment by KBP-TDC** [comment]
@linehrr Good feedback! We do have plans to improve it. One challenge is to expose to much of implementation details and should not be user-visible. We will try to find the right balance (our Tuning section https://prestodb.io/docs/current/admin/tuning.html is definitely not enough).

**Comment by linehrr** [comment]
adding parameter:
> exchange.http-client.request-timeout=120s

does help us on the query successful rate.
our cluster spec:
> 114 nodes (Dell R410/R320)
> 16GB/28GB/32GB RAM config
> mapred/Presto hybrid, parallel running

less crash is observed after setting change.

**Comment by hustnn** [comment]
@linehrr 

Is parameter "exchange.http-client.request-timeout" still working now? I can't find it in 0.194 source code. 

What is the difference between "scheduler.http-client.request-timeout" and "exchange.http-client.request-timeout" here?

**Comment by linehrr** [comment]
@hustnn sorry, I lost track of presto releases. we stopped using Presto a while ago and moved to sparkSQL. I think they refactored code base quite a lot. so please find the latest docs or latest source code.

**Comment by hustnn** [comment]
@linehrr I see. Thanks all the same.
